### PR TITLE
Instead of the clock algorithm, always evict the largest layer.

### PR DESCRIPTION
**Pull request by hlinnaka**
This avoids nasty behavior where one large layer would hog all the
memory, causing many many other small layers to be evicted, which leads
to write amplification and bad performance. That was quite visible in
a plain pgbench workload.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Before this, pgbench with scale factor 100 performed very badly:
```
./tmp_install/bin/pgbench -c5  -P1 -T2500 -N postgres
starting vacuum...end.
progress: 1.0 s, 852.8 tps, lat 4.997 ms stddev 50.638
progress: 2.0 s, 52.0 tps, lat 100.735 ms stddev 222.197
progress: 3.0 s, 26.0 tps, lat 149.496 ms stddev 304.346
progress: 4.0 s, 46.0 tps, lat 102.143 ms stddev 195.050
progress: 5.0 s, 35.0 tps, lat 169.081 ms stddev 378.483
progress: 6.0 s, 32.0 tps, lat 87.653 ms stddev 170.398
progress: 7.0 s, 52.0 tps, lat 138.552 ms stddev 348.392
progress: 8.0 s, 59.0 tps, lat 94.560 ms stddev 229.055
progress: 9.0 s, 98.0 tps, lat 52.913 ms stddev 83.760
progress: 10.0 s, 48.0 tps, lat 59.247 ms stddev 107.298
progress: 11.0 s, 45.0 tps, lat 161.811 ms stddev 376.813
progress: 12.0 s, 101.0 tps, lat 41.897 ms stddev 118.050
progress: 13.0 s, 128.0 tps, lat 35.842 ms stddev 79.357
progress: 14.0 s, 11.0 tps, lat 564.068 ms stddev 565.282
progress: 15.0 s, 122.9 tps, lat 30.978 ms stddev 76.477
progress: 16.0 s, 66.0 tps, lat 85.322 ms stddev 145.875
progress: 17.0 s, 38.0 tps, lat 118.411 ms stddev 245.746
progress: 18.0 s, 100.0 tps, lat 53.310 ms stddev 130.407
progress: 19.0 s, 74.0 tps, lat 56.294 ms stddev 146.922
progress: 20.0 s, 46.0 tps, lat 114.777 ms stddev 226.828
...
```
What happens is that the all the memory gets hogged by one layer. As long as there is a moderate amount of access to that layer, its usage count stays high, and it stays in memory, and there is very little memory left for other layers. As a result, in-memory layers for other segments keep getting created and quickly evicted in a tight loop.

There are a few things we could do here:
1. Make the eviction algorithm smarter, preferring to evict layers with a larger footprint first. Evicting small layers is bad, because it leads to a lot of small files, and write amplification because we always materialize a new image layer when a layer is evicted. But once a layer is "large enough", you reach a point of diminishing returns from prolonging its eviction. I have no numbers for what is "small", "large", or "large enough" here, but that's the general tendency.

2. Make evicting a small layer cheaper. The expensive part is materializing a new image layer, but we could skip that in many cases, and only write a new image layer if there has been enough churn on the segment since the last image. No need to do it after each delta layer. Another aspect of small layers is that you end up with a lot of small files; that doesn't show up in pgbench so easily, but there is some overhead in tracking each layer, and dealing with a lot of files in S3 is slow and expensive.

3. To addressing the problem of small files, we could merge consecutive delta files.

This PR does 1. The new algorithm is very straightforward: we always evict the in-memory layer that consumes the most memory. I'm sure it's not optimal, but it's simple, and works pretty well for the pgbench case at least. The worst case scenario for this algorithm would be something like this: Create a lot of small tables, and insert a few rows into each, so that the in-memory layers for those tables use up most of the memory. Then create one more table, and perform a lot of updates on it. The layer for that one table will grow to be only a little bit larger than any of the small table layers, but still very small, until the memory is full. Then the system will evict that layer, because it's larger than the others (but still small). And repeat. The checkpoint distance puts limit on how bad that can become, eventually the small tables will become old enough that they will be evicted by the checkpointer, freeing up the memory. But you can then recreate the problem by making another round if inserts to each of the small tables.

Anyway, I think this is better than what we had. We should continue optimizing the algorithm, and also do 2-3, but this fixes the most obvious bad case with pgbench.

After this PR, pgbench output looks like this:
```
starting vacuum...end.
progress: 1.0 s, 3865.5 tps, lat 1.233 ms stddev 2.160
progress: 2.0 s, 3839.5 tps, lat 1.337 ms stddev 2.786
progress: 3.0 s, 2627.5 tps, lat 1.823 ms stddev 4.264
progress: 4.0 s, 2639.3 tps, lat 1.964 ms stddev 5.371
progress: 5.0 s, 2550.0 tps, lat 1.951 ms stddev 7.844
progress: 6.0 s, 1329.9 tps, lat 3.637 ms stddev 8.438
progress: 7.0 s, 3932.5 tps, lat 1.310 ms stddev 2.545
progress: 8.0 s, 2861.0 tps, lat 1.742 ms stddev 6.565
progress: 9.0 s, 4723.8 tps, lat 1.054 ms stddev 0.539
progress: 10.0 s, 4017.1 tps, lat 1.237 ms stddev 0.908
progress: 11.0 s, 3730.8 tps, lat 1.337 ms stddev 1.385
progress: 12.0 s, 3499.0 tps, lat 1.423 ms stddev 1.858
progress: 13.0 s, 2850.1 tps, lat 1.751 ms stddev 3.963
progress: 14.0 s, 3760.8 tps, lat 1.320 ms stddev 1.401
progress: 15.0 s, 2386.0 tps, lat 2.097 ms stddev 6.026
progress: 16.0 s, 2841.0 tps, lat 1.749 ms stddev 4.299
progress: 17.0 s, 2371.1 tps, lat 2.112 ms stddev 5.723
progress: 18.0 s, 2807.1 tps, lat 1.776 ms stddev 4.323
progress: 19.0 s, 1866.0 tps, lat 2.676 ms stddev 11.860
progress: 20.0 s, 2549.8 tps, lat 1.947 ms stddev 4.789
...
```

**Review by knizhnik — APPROVED** [review]
Certainly such speed improvement looks very promising.
But I am not sure that this strategy always leads to the best results.
Well, I certainly realize that none strategy can be optima in all cases.
But let's consider quite common case: we have populated N table and start insertion in N+1 table. We have open layers for N relation and start filling layer for N+1 relation.Once it becomes the largest layer, we will evict it before it is completely filled. We do not need to keep open other layers, because we already done with this relations. But the eviction algorithm knows nothing about it. And even the fact that them are not updated for a long time doesn't help it.

So, it just once again proves that it is much more difficult to choose efficient strategy for layered storage which will provide even not optimal, but not ugly behavior for most common access patterns:(

Some more questions which are not so important:
1. Why you preserved access counter if it is not used now?
2. Why you have used clock algorithm, and not LRU for example? The main advantage of clock is that it doesn't require global lock. But it seems to me that we are in any case executing this code under mutex. And in all other aspects LRU is better: is faster and provides more optimal behavior.

File: pageserver/src/layered_repository/global_layer_map.rs [hunk 1]
```diff
@@ -49,14 +51,17 @@ enum SlotData {
 struct Slot {
     tag: u64,
     data: SlotData,
-    usage_count: AtomicU8, // for clock algorithm
+    /// Tracks how frequently this layer has been accessed recently. TODO: This is
+    /// intended for a clock algorithm, but currently the eviction is based purely
+    /// on memory usage, not how recently a layer has been accessed. So currently
+    /// this is unused.
+    usage_count: AtomicU8,
```
Comment (patins) [inline pageserver/src/layered_repository/global_layer_map.rs:58]: Completely cut this out before commit?

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 2]
```diff
@@ -310,6 +303,24 @@ impl InMemoryLayer {
...
     }
 
+    /// Callback to notify when this layer is inserted to the global layer map.
+    /// From this point on, we need to keep the global layer map informed whenever
+    /// the memory consumption of this layer increases, by calling
+    /// [`update_layer_memory_stats`].
+    pub fn register_layer_id(&self, layer_id: LayerId) {
+        let mut inner = self.inner.write().unwrap();
+
+        inner.layer_id = Some(layer_id);
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:313]: Should layer id always be None in this case

File: pageserver/src/layered_repository/global_layer_map.rs [hunk 3]
```diff
@@ -137,50 +146,47 @@ impl OpenLayers {
...
+        for i in 0..self.slots.len() {
+            let slot = &self.slots[i];
+
+            if let SlotData::Occupied(_data, mem_usage) = &slot.data {
+                if *mem_usage > best_mem_usage {
+                    best_slot = i;
+                    best_mem_usage = *mem_usage;
                 }
+            }
+        }
```
Comment (patins) [inline pageserver/src/layered_repository/global_layer_map.rs:173]: Consider some `iter` `enumerate` `max_by_key` type thing

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 4]
```diff
@@ -310,6 +303,24 @@ impl InMemoryLayer {
...
+    }
+
+    /// Callback from the global layer map, when this layer is removed from global
+    /// layer map.
+    pub fn unregister(&self) {
+        let mut inner = self.inner.write().unwrap();
+
+        inner.layer_id = None;
+    }
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:322]: I don't think this is bad, but just wondering will we ever remove a layer and then add it back to the global layer map?

**Comment by hlinnaka** [comment]
This is no longer relevant after #864, closing

**Pull request closed without merging**